src/include/mallocMC/creationPolicies/Scatter.hpp

```diff
@@ -33,1342 +27,465 @@
-            }
+        ALPAKA_FN_ACC static auto startIndex(auto const& acc)
+        {
+            return (laneid() * alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc).sum()) % numPages();
+        }
 
-        public:
-            /**
-             * create allocates the requested number of bytes via the heap.
-             * Coalescing has to be done before by another policy.
-             * @param bytes number of bytes to allocate
-             * @return pointer to the allocated memory
-             */
-            template<typename AlignmentPolicy, typename AlpakaAcc>
-            ALPAKA_FN_ACC auto create(const AlpakaAcc& acc, uint32 bytes) -> void*
+        template<typename TAcc>
+        ALPAKA_FN_ACC auto createChunk(TAcc const& acc, uint32_t const numBytes) -> void*
+        {
+            auto index = startIndex(acc);
+
+            // Under high pressure, this loop could potentially run for a long time because the information where and
+            // when we started our search is not maintained and/or used. This is a feature, not a bug: Given a
+            // consistent state, the loop will terminate once a free chunk is found or when all chunks are filled for
+            // long enough that `choosePage` could verify that each page is filled in a single run.
+            //
+            // The seemingly non-terminating behaviour that we wrap around multiple times can only occur (assuming a
+            // consistent, valid state of the data) when there is high demand for memory such that pages that appear
+            // free to `choosePage` are repeatedly found but then the free chunks are scooped away by other threads.
+            //
+            // In the latter case, it is considered desirable to wrap around multiple times until the thread was fast
+            // enough to acquire some memory.
+            index = choosePage(acc, numBytes, index);
+            void* pointer = index != noFreePageFound()
+                ? PageInterpretation<T_pageSize>{pages[index], numBytes}.create(acc)
+                : nullptr;
+
+            while(index != noFreePageFound() and pointer == nullptr)
             {
-                if(bytes == 0)
-                    return 0;
-                /* Take care of padding
-                 * bytes = (bytes + dataAlignment - 1) & ~(dataAlignment-1);
-                 * in alignment-policy.
-                 * bytes == pagesize must be handled by allocChunked() else maxchunksize calculation based
-                 * on the waste factor is colliding with the allocation schema in allocPageBased().
-                 */
-                if(bytes <= pagesize)
-                    // chunck based
-                    return allocChunked<AlignmentPolicy>(acc, bytes);
-                else
-                    // allocate a range of pages
-                    return allocPageBased(acc, bytes);
+                leavePage(acc, index);
+                ++index;
```

Comment:
BUG: must be `index = (index + 1u) % numPages();` else index can become to large

Reply:
Nope, that's not a problem:
1. `index` can get out of range (slightly, i.e., by 1 if `index` was pointing to the last page before) but that's not a problem because it is only used in a `wrappingLoop`. Any `startIndex` out of range will simply result in effectively starting from 0.
2. `index` is set on the following line to the new hoped-to-be-free page which is again a valid (potentially wrapped around) index in range, so `index` cannot become arbitrarily large.